### PR TITLE
refactor(tidy): Use C++20 contains method

### PR DESCRIPTION
**Pull request by aureleoules**
C++20 introduced the `contains` method on containers to check if an element is in the container. It can be used instead of the `count` method now.
I believe it is easier to understand, as `contains` directly returns a `bool` indicating whether the element exists, while `count` returns the number of occurrences of the element, which then often needs to be compared against 0.
Also, it is slightly more efficient than count for this use-case.

This pull request introduces the clang-tidy check `readability-container-contains` and fixes the instances where `count` is being used where `contains` could be used instead. I used the clang-tidy -fix option to do it automatically.

### CODE REVIEW

**Comment by theuni** [comment]
Nice use of tidy!

Concept ACK, assuming `::contains()` exists for common stdlibs by now. Waiting to see what c-i thinks.

**Comment by TheCharlatan** [comment]
Concept ACK

**Review by BrandonOdiwuor — COMMENTED** [review]
Concept ACK using ::contains(Key& key) (C++20) to check for elements

**Comment by TheCharlatan** [comment]
Can the full diff be reproduced with `run-clang-tidy -fix`? If so, can you post the full commands you are using? I am getting a slightly different diff if I run it myself.

**Comment by aureleoules** [comment]
> Can the full diff be reproduced with run-clang-tidy -fix? If so, can you post the full commands you are using? I am getting a slightly different diff if I run it myself.

Sorry, my initial push did not pass the tidy check because it seems the `-fix` option does not fix code in macros (all the asserts, Assume, etc.) so I manually fixed these in a second push. This may be the reason why you are not getting the same diff as me.

I don't have the specific command I used on hand but I used the `find` command to retrieve all cpp and h files, piped to `clang-tidy-17 -fix` if I remember correctly.

**Comment by TheCharlatan** [comment]
Re https://github.com/bitcoin/bitcoin/pull/29191#issuecomment-1880843285

> Sorry, my initial push did not pass the tidy check because it seems the -fix option does not fix code in macros (all the asserts, Assume, etc.) so I manually fixed these in a second push. This may be the reason why you are not getting the same diff as me.

I see, that is a bit unfortunate. I also found some more things to change: https://github.com/TheCharlatan/bitcoin/commit/63fcfbecd133551032857187e5e639d36a2f7b06, but not sure if non-tidy changes should be made here tbh.

**Comment by aureleoules** [comment]
> I see, that is a bit unfortunate. I also found some more things to change: https://github.com/TheCharlatan/bitcoin/commit/63fcfbecd133551032857187e5e639d36a2f7b06, but not sure if non-tidy changes should be made here tbh.

Thanks for the commit, I added it. Did you find them manually or with clang-tidy? 
I'm not sure why the CI didn't catch them...

File: src/addrman.cpp [hunk 1]
```diff
@@ -909,7 +909,7 @@ void AddrManImpl::ResolveCollisions_()
         bool erase_collision = false;
 
         // If id_new not found in mapInfo remove it from m_tried_collisions
-        if (mapInfo.count(id_new) != 1) {
+        if (!mapInfo.count(id_new)) {
```
Comment (mzumsande) [inline src/addrman.cpp:912]: `count` -> `contains`

Comment (aureleoules) [reply src/addrman.cpp]: Thanks, fixed.

**Comment by stickies-v** [comment]
I support using `contains` going forward, but I'm not sure for this PR the benefits outweigh the costs?
1. I prefer the `contains` interface but imo using `count` is also pretty straightforward, so the benefits seem limited?
2. It conflicts with a large number of PRs
3. It's not a scripted diff

I think I'd prefer removing the clang-tidy check for now and only changing it in places where it doesn't conflict with other PRs, e.g. maybe in tests? And then organically have more `contains` adoption over time?

**Comment by TheCharlatan** [comment]
Re https://github.com/bitcoin/bitcoin/pull/29191#issuecomment-1882971897

>  but I'm not sure for this PR the benefits outweigh the costs?

FWIW I would approve this PR if it would only contain the changes that are automatically picked up by tidy and applied by running with `-fix`, making resolving potential conflicts purely manual.

**Comment by aureleoules** [comment]
> FWIW I would approve this PR if it would only contain the changes that are automatically picked up by tidy and applied by running with -fix, making resolving potential conflicts purely manual.

Sure, I agree. Closing this pull.

**Pull request closed without merging**